### PR TITLE
Testing Trainer support for multiple artifact input with LatestArtifactResolver ResolverNode

### PR DESCRIPTION
**Pull request by 18jeffreyma**
Testing ResolverNode's output of a channel containing multiple `Examples` artifacts feeding into `Trainer` component. Test file is titled `iris_resolver_e2e_test.py`.

### CODE REVIEW

File: tfx/components/trainer/fn_args_utils.py [hunk 1]
```diff
@@ -59,14 +59,14 @@ def get_common_fn_args(input_dict: Dict[Text, List[types.Artifact]],
                        working_dir: Text = None) -> FnArgs:
   """Get common args of training and tuning."""
   train_files = [
-      io_utils.all_files_pattern(
-          artifact_utils.get_split_uri(input_dict[constants.EXAMPLES_KEY],
-                                       'train'))
+      io_utils.all_files_pattern(uri) for uri in
+      artifact_utils.get_split_uris(input_dict[constants.EXAMPLES_KEY],
```
Comment (1025KB) [inline tfx/components/trainer/fn_args_utils.py:63]: Overall lgtm, could you print `input_dict[constants.EXAMPLES_KEY]` here, and paste result here? I want to check the uri and id to see if it matches what examplegen generates

File: tfx/examples/iris/iris_resolver_e2e_test.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      instance_name='schema_resolver',
+      resolver_class=latest_artifacts_resolver.LatestArtifactsResolver,
+      schema=Channel(type=Schema))
+
+  # Resolve latest two example artifacts into one channel for trainer.
+  latest_examples_resolver = ResolverNode(
+      instance_name='latest_examples_resolver',
+      resolver_class=latest_artifacts_resolver.LatestArtifactsResolver,
+      resolver_configs={'desired_num_of_artifacts': window_size},
+      latest_n_examples=Channel(type=Examples))
```
Comment (1025KB) [inline tfx/examples/iris/iris_resolver_e2e_test.py:99]: could you verify the if we put examplegen and resolver in the same pipeline, we can make sure resolver depends on examplegen

window size 3:
pipeline1: examplegen -> tdfv
pipeline2:examplegen ->tfdv
pipeline3: examplegen->tfdv ->resolver->trainer

File: tfx/components/trainer/fn_args_utils.py (hunk 1 above)
Comment (18jeffreyma) [reply tfx/components/trainer/fn_args_utils.py:63]: Attached is that specific log:

`
0717 20:04:34.873566 135202664122112 fn_args_utils.py:63] [Artifact(artifact: id: 1
type_id: 5
uri: "/tmp/iris_resolver_e2e_testfc2pebe5/tmpa0t72p3g/testIrisPipelineResolverWithDependency/tfx/pipelines/resolver_test/CsvExampleGen/examples/1"
properties {
  key: "split_names"
  value {
    string_value: "[\"train\", \"eval\"]"
  }
}
custom_properties {
  key: "input_fingerprint"
  value {
    string_value: "split:single_split,num_files:1,total_bytes:2758,xor_checksum:1595041459,sum_checksum:1595041459"
  }
}
custom_properties {
  key: "name"
  value {
    string_value: "examples"
  }
}
custom_properties {
  key: "payload_format"
  value {
    string_value: "FORMAT_TF_EXAMPLE"
  }
}
custom_properties {
  key: "pipeline_name"
  value {
    string_value: "resolver_test"
  }
}
custom_properties {
  key: "producer_component"
  value {
    string_value: "CsvExampleGen"
  }
}
custom_properties {
  key: "span"
  value {
    string_value: "0"
  }
}
custom_properties {
  key: "state"
  value {
    string_value: "published"
  }
}
create_time_since_epoch: 1595041459439
last_update_time_since_epoch: 1595041460679
, artifact_type: id: 5
name: "Examples"
properties {
  key: "span"
  value: INT
}
properties {
  key: "split_names"
  value: STRING
}
), Artifact(artifact: id: 4
type_id: 5
uri: "/tmp/iris_resolver_e2e_testfc2pebe5/tmpa0t72p3g/testIrisPipelineResolverWithDependency/tfx/pipelines/resolver_test/CsvExampleGen/examples/4"
properties {
  key: "split_names"
  value {
    string_value: "[\"train\", \"eval\"]"
  }
}
custom_properties {
  key: "input_fingerprint"
  value {
    string_value: "split:single_split,num_files:1,total_bytes:2758,xor_checksum:1595041464,sum_checksum:1595041464"
  }
}
custom_properties {
  key: "name"
  value {
    string_value: "examples"
  }
}
custom_properties {
  key: "payload_format"
  value {
    string_value: "FORMAT_TF_EXAMPLE"
  }
}
custom_properties {
  key: "pipeline_name"
  value {
    string_value: "resolver_test"
  }
}
custom_properties {
  key: "producer_component"
  value {
    string_value: "CsvExampleGen"
  }
}
custom_properties {
  key: "span"
  value {
    string_value: "1"
  }
}
custom_properties {
  key: "state"
  value {
    string_value: "published"
  }
}
create_time_since_epoch: 1595041464576
last_update_time_since_epoch: 1595041465895
, artifact_type: id: 5
name: "Examples"
properties {
  key: "span"
  value: INT
}
properties {
  key: "split_names"
  value: STRING
}
), Artifact(artifact: id: 7
type_id: 5
uri: "/tmp/iris_resolver_e2e_testfc2pebe5/tmpa0t72p3g/testIrisPipelineResolverWithDependency/tfx/pipelines/resolver_test/CsvExampleGen/examples/7"
properties {
  key: "split_names"
  value {
    string_value: "[\"train\", \"eval\"]"
  }
}
custom_properties {
  key: "input_fingerprint"
  value {
    string_value: "split:single_split,num_files:1,total_bytes:2758,xor_checksum:1595041469,sum_checksum:1595041469"
  }
}
custom_properties {
  key: "name"
  value {
    string_value: "examples"
  }
}
custom_properties {
  key: "payload_format"
  value {
    string_value: "FORMAT_TF_EXAMPLE"
  }
}
custom_properties {
  key: "pipeline_name"
  value {
    string_value: "resolver_test"
  }
}
custom_properties {
  key: "producer_component"
  value {
    string_value: "CsvExampleGen"
  }
}
custom_properties {
  key: "span"
  value {
    string_value: "2"
  }
}
custom_properties {
  key: "state"
  value {
    string_value: "published"
  }
}
create_time_since_epoch: 1595041469775
last_update_time_since_epoch: 1595041471049
, artifact_type: id: 5
name: "Examples"
properties {
  key: "span"
  value: INT
}
properties {
  key: "split_names"
  value: STRING
}
)]`

File: tfx/examples/iris/iris_resolver_e2e_test.py (hunk 2 above)
Comment (18jeffreyma) [reply tfx/examples/iris/iris_resolver_e2e_test.py:99]: Updated test to also test case for dependency.

**Comment by 18jeffreyma** [comment]
Resolved above feedback, thanks!

File: tfx/examples/iris/iris_resolver_e2e_test.py (hunk 2 above)
Comment (18jeffreyma) [reply tfx/examples/iris/iris_resolver_e2e_test.py:97]: From running this "full pipeline" and printing out the dependency graph, we get the following:

`I0720 13:11:47.808361 135087534106432 beam_dag_runner.py:151] Component CsvExampleGen depends on [].
I0720 13:11:47.829361 135087534106432 beam_dag_runner.py:165] Component CsvExampleGen is scheduled.
I0720 13:11:47.836360 135087534106432 beam_dag_runner.py:151] Component ResolverNode.latest_examples_resolver depends on ['Run[CsvExampleGen]'].
I0720 13:11:47.836360 135087534106432 beam_dag_runner.py:165] Component ResolverNode.latest_examples_resolver is scheduled.
I0720 13:11:47.843361 135087534106432 beam_dag_runner.py:151] Component StatisticsGen depends on ['Run[CsvExampleGen]'].
I0720 13:11:47.844361 135087534106432 beam_dag_runner.py:165] Component StatisticsGen is scheduled.
I0720 13:11:47.844361 135087534106432 beam_dag_runner.py:151] Component SchemaGen depends on ['Run[StatisticsGen]'].
I0720 13:11:47.851361 135087534106432 beam_dag_runner.py:165] Component SchemaGen is scheduled.
I0720 13:11:47.858361 135087534106432 beam_dag_runner.py:151] Component Trainer depends on ['Run[SchemaGen]', 'Run[ResolverNode.latest_examples_resolver]'].
I0720 13:11:47.858361 135087534106432 beam_dag_runner.py:165] Component Trainer is scheduled.`

**Comment by 18jeffreyma** [comment]
Added note which contains dependency graph for the full pipeline. ExampleGen is explicitly added as a dependency for ResolverNode.

File: tfx/examples/iris/iris_resolver_e2e_test.py (hunk 2 above)
Comment (1025KB) [reply tfx/examples/iris/iris_resolver_e2e_test.py:97]: nice!

**Pull request closed without merging**